### ISSUE TITLE
Add a simple way to pass context to the Conversation service

### ISSUE THREAD

**Issue by trpmap93** [watson-developer-cloud/swift-sdk#598]
Remember, an issue is not the place to ask questions. You can use [Stack Overflow](http://stackoverflow.com/questions/tagged/ibm-watson) for that, or you may want to start a discussion on the [dW Answers](https://developer.ibm.com/answers/questions/ask/?topics=watson).

**Comment by trpmap93** [comment]
There is no documentation on how to pass context using the Swift SDK that I can find.  Below is a sample of code that we figured out over a 2 month process to be able to do so:

  let failure = { (error: Error) in print(error) }
        conversation.message(withWorkspace: conWorkspace, failure: failure) { response in
            self.context = response.context
            
            // Grab the Conversation ID and System object and copy their contents
            let convID = (self.context?.conversationID)!
            let systm = (self.context?.system)!

            let jason = JSON(json: ["conversation_id": convID, "system": systm.json, "json": [["mrn":"MP0312"], ["cid":"PDX"], ["prov":"MP"]]])
                 
            //Now re-initialize the context object with the json object you created above
            self.context = try! Context(json: jason)
            
            // Form new message request containing your initial context variable(s) and pass it to the Watson Conversation service
            let conFailure = { (error: Error) in print(error)}
            let request = MessageRequest(text: "hi", context: (self.context))
            self.conversation.message(withWorkspace: self.conWorkspace, request: request, failure: conFailure) {
                response in
                
                //copies the response context to context
                self.context = response.context

...
This took a lot of reverse engineering to figure out so there must be a more simple way.  In addition if you use my  solution above when you change this is the Conversation service dialog, you have to updates the context as so:

 "context": {
    "json": {
      "mrn": "123456",
      "cid": "BNA",
      "prov": "RM"
    }

**Comment by germanattanasio** [comment]
@schen22 can you take a look at this? I met @trpmap93 at Interconnect and the had some issues trying to use the SDK.

**Comment by schen22** [comment]
Hi @trpmap93 - thanks for letting us know. We actually have an example in our README about how to [continue a conversation](https://github.com/watson-developer-cloud/swift-sdk#conversation) which includes an example about how to pass and use the context. If you have any suggestions on where else we can place this information, or how to improve on our explanation so that other users will more easily understand how to use the Conversation service, definitely feel free to let us know. At the same time I agree the example in the README is fairly simple, and perhaps not the best fit for what you were looking for. 

Are you trying to customize the responses the Conversation service gives, or are you trying to hold a conversation with the Conversation? 

From your code it looks like you're not changing or updating the `Context` variable after you instantiate it with the `JSON` object you created, so I'll go under the assumption you're trying to hold a conversation with the service. 

I'll walk through the code you posted. Let me know if you agree/disagree:
```  let failure = { (error: Error) in print(error) }
        conversation.message(withWorkspace: conWorkspace, failure: failure) { response in
            self.context = response.context
 ```
You save the context here, but you end up creating your own context. Why? You're basically copying what the `Context` object contains again in the `jason` object you create in the following
```
            // Grab the Conversation ID and System object and copy their contents
            let convID = (self.context?.conversationID)!
            let systm = (self.context?.system)!

            let jason = JSON(json: ["conversation_id": convID, "system": systm.json, "json": [["mrn":"MP0312"], ["cid":"PDX"], ["prov":"MP"]]])
                 
            //Now re-initialize the context object with the json object you created above
            self.context = try! Context(json: jason)
```
Remove all the code above. You needn't re-initialize the context object since you've already saved it in the line `self.context = response.context`. Furthermore, in your second callback you actually do what I'm suggesting. Though maybe in the rest of the code you create another JSON object. My point is, creating the `jason` object you made is redundant. Just pass in the `response.context` variable you received from the conversation service.

Simplify your code to just the following:
```
  let failure = { (error: Error) in print(error) }
        conversation.message(withWorkspace: conWorkspace, failure: failure) { response in
            self.context = response.context
            // Form new message request containing your initial context variable(s) and pass it to the Watson Conversation service
            let conFailure = { (error: Error) in print(error)}
            let request = MessageRequest(text: "hi", context: (self.context))
            self.conversation.message(withWorkspace: self.conWorkspace, request: request, failure: conFailure) {
                response in
                
                //copies the response context to context
                self.context = response.context

```

From your code it looks like the main difficulty was how to handle callbacks. All calls to the Watson services are done through HTTP requests that are asynchronous meaning we need to wait for the service's response through callbacks. There's no way around the asynchronous calls, but maybe adding a completion handler when using the SDK would help you. That way you don't need to have a callback within a callback. 

Definitely let us know what was hard to understand so we can improve. No one should spend 2 months or more even trying to figure one thing out. Try reaching out to us sooner too. Just keep in mind limited time and resources will lead to prolonged delays so sometimes you'll have to keep following up with any open issues you create with us unfortunately. However overall our main goal is to improve the ease of using the SDK, so we want to help.

**Comment by schen22** [comment]
You can also check out [Cognitive Concierge](https://github.com/IBM-MIL/CognitiveConcierge/blob/master/CognitiveConcierge-iOS/CognitiveConcierge/MessagesViewModel.swift#L60), a sample app we wrote that actually uses the Conversation service in a full fledged app. The context is passed through a segue in the EntertainmentViewController.swift file, to the MessagesViewModel.swift, which is then called in the MessagesViewController.swift to handle the conversation response, etc.

**Comment by trpmap93** [comment]
Hi Schen,  I guess a better way to describe what I was trying to do was actually set a context variable from my application and pass this to the service.  I cannot find any example of how to do this using the sdk.  Basically what I was wanting to do was send out the nil call and receive back the conversation id and system object.  I then wanted to pass those along with an initial context variable object that I define in my application to the conversation service.  I can then modify this context variable in the conversation service and when need be alter it in my application code to pass back.  

Hopefully this makes sense :)

And thank you for your detail response above!

**Comment by trpmap93** [comment]
The example I gave above was just to set the context variable initially when the service is started.  I have another function that handles the conversation back and forth after the initializing function above.  Its just imperative with application I am building to be able to set some initial items in the context variable from the application side. 

Thanks again!

**Comment by schen22** [comment]
Oh I see - so it's mainly difficulty initializing the service then. Have you tried using completion handlers? Creating a context variable isn't necessary in my opinion, and having completion handlers is a bit clearer to understand, though that's all personal preference. It also makes more sense to contact the Conversation service to grab the Context variable rather than have the SDK provide a pre-defined Context variable to pass to the service for users to use. I'm not sure how else to better design creating/beginning a conversation. Any suggestions or thoughts you have would be much appreciated!

Let me know what you think about the following example that describes what I'm thinking:

```
let failure = { (error: Error) in print(error) }

// Instantiate context object outside of callback in order to continue conversation outside of the callback to call on and continue modifying the variable.
var context: Context?
```

Here I use completion handlers to instantiate the conversation service and upon success, return the needed context to continue a conversation.
```
func instantiateConversationContext(
    onSuccess success: @escaping(Context) -> Void, 
    onFailure failure: @escaping (Error) -> Void) {
    conversation.message(withWorkspace: conWorkspace, failure: failure) { response in
        // End callback with success call that passes context variable
        success(response.context)
        // end this callback to improve code readability
  }
}
```

The following provides an example of how to use the instantiateConversationContext function created. Rather than initializing and creating your own context object, we call on the function to grab the context to use in the conversation service. The rest of my example is pretty dependent on what your app needs; I just added it to describe how to use the function I created above.

```
// Call on the instantiateConversationContext func to grab the context.
instantiateConversationContext(
  onSuccess: {
    // save context as a variable 
    self.context = response.context

    // Add your app logic here. 

    // Begin or continue conversation. 
    continueConversation(message: "hi") { conversationReply in
      // Or whatever you want to do with the reply.
      print (conversationReply)
    }
  },
  onFailure: { error in 
    print (error)
  }
)  
```

The above continueConversation call is a function I made that takes in whatever message you want to send to the service and outputs a completion handler for you to check what the response is:

```
func continueConversation(message: String, completion: @escaping ([String]) -> Void) {
  // Form new message request as you had done.
  let conFailure = { (error: Error) in print(error)}
  let request = MessageRequest(text: message, context: (self.context))
  // Pass into the conversation service
  conversation.message(withWorkspace: self.conWorkspace, request: request, failure: conFailure) {
    response in
      self.context = response.context
      
      // output from conversation is usually saved as well so we can pass it into 
      // the completion.
      completion(response.output.text)
  }
}
```

**Comment by trpmap93** [comment]
Schen22 thanks again for the example code.  At issue is still an easy way to change the context variable from the application side.  As I understand, context is made up of an conversation_id, a system object, and a json object (which is the actual context variable part).  The example I gave at the very first will work but instead of the context variable being expressed as context:{"property_one_key": "property_one_value" ...   it would have to be expressed as context: {json: {"property_one_key": "property_one_value"...
Changing the context variable looks pretty trivial in other languages from the application side, but I can't see a way to do it other than how I suggested which seems rather complex.
Thanks again for the discussion.
